streams/shared/src/main/scala/scalaz/stream/ZStream.scala

```diff
@@ -18,52 +18,81 @@ package scalaz.zio.stream
- *
- */
-trait ZStream[-R, +E, +A] extends Serializable { self =>
+trait ZStream[-R, +E, +A] { self =>
   import ZStream.Fold
 
-  /**
-   * Executes an effectful fold over the stream of values.
-   */
   def fold[R1 <: R, E1 >: E, A1 >: A, S]: Fold[R1, E1, A1, S]
 
   def foldLeft[A1 >: A, S](s: S)(f: (S, A1) => S): ZIO[R, E, S] =
-    self.fold[R, E, A1, S].flatMap(f0 => f0(s, _ => true, (s, a) => IO.succeed(f(s, a))))
+    fold[R, E, A1, S].flatMap(fold => fold(s, _ => true, (s, a) => ZIO.succeed(f(s, a))).use(ZIO.succeed))
+
+  final def ++[R1 <: R, E1 >: E, A1 >: A](other: ZStream[R1, E1, A1]): ZStream[R1, E1, A1] =
+    new ZStream[R1, E1, A1] {
+      def fold[R2 <: R1, E2 >: E1, A2 >: A1, S]: Fold[R2, E2, A2, S] =
+        ZIO.succeedLazy { (s, cont, f) =>
+          ZManaged.unwrap {
+            self.fold[R2, E2, A2, S].map { foldLeft =>
+              foldLeft(s, cont, f).flatMap { s =>
+                if (!cont(s)) ZManaged.succeed(s)
+                else
+                  ZManaged.unwrap {
+                    other.fold[R2, E2, A2, S].map(foldRight => foldRight(s, cont, f))
+                  }
+              }
+            }
+          }
+        }
+    }
 
   /**
-   * Concatenates the specified stream to this stream.
+   * Allow a faster producer to progress independently of a slower consumer by buffering
+   * up to `capacity` elements in a queue.
+   *
+   * @note when possible, prefer capacities that are powers of 2 for better performance.
    */
-  final def ++[R1 <: R, E1 >: E, A1 >: A](that: => ZStream[R1, E1, A1]): ZStream[R1, E1, A1] =
-    new ZStream[R1, E1, A1] {
-      override def fold[R2 <: R1, E2 >: E1, A2 >: A1, S]: Fold[R2, E2, A2, S] =
+  final def buffer(capacity: Int): ZStream[R, E, A] =
+    ZStream.managed(self.toQueue(capacity)).flatMap { queue =>
+      ZStream.fromQueue(queue).unTake
+    }
+
+  /**
+   * Performs a filter and map in a single step.
+   */
+  def collect[B](pf: PartialFunction[A, B]): ZStream[R, E, B] =
+    new ZStream[R, E, B] {
+      override def fold[R1 <: R, E1 >: E, B1 >: B, S]: Fold[R1, E1, B1, S] =
         IO.succeedLazy { (s, cont, f) =>
-          self.fold[R2, E2, A, S].flatMap { f0 =>
-            f0(s, cont, f).flatMap { s =>
-              if (cont(s)) that.fold[R2, E2, A1, S].flatMap(f1 => f1(s, cont, f))
-              else IO.succeed(s)
+          ZManaged.unwrap {
+            self.fold[R1, E1, A, S].map { fold =>
+              fold(s, cont, (s, a) => if (pf.isDefinedAt(a)) f(s, pf(a)) else IO.succeed(s))
             }
           }
         }
     }
 
+  /**
+   * Transforms all elements of the stream for as long as the specified partial function is defined.
+   */
+  def collectWhile[B](pred: PartialFunction[A, B]): ZStream[R, E, B] = new ZStream[R, E, B] {
+    override def fold[R1 <: R, E1 >: E, B1 >: B, S]: Fold[R1, E1, B1, S] =
+      IO.succeedLazy { (s, cont, f) =>
+        ZManaged.unwrap {
```

Comment:
@jdegoes `ZManaged.unwrap { fold.map(...) }` repeats itself all over the file which is unfortunate

Reply:
I saw that. Any way to factor out that duplication?

Reply:
We could have this as a private helper:
```
  private def foldUnwrap[R1 <: R, E1 >: E, A1 >: A, S] =
    ZManaged.fromEffect(self.fold[R1, E1, A1, S])
```

And then the combinators look like this:
```
  final def ++[R1 <: R, E1 >: E, A1 >: A](other: ZStream[R1, E1, A1]): ZStream[R1, E1, A1] =
    new ZStream[R1, E1, A1] {
      def fold[R2 <: R1, E2 >: E1, A2 >: A1, S]: Fold[R2, E2, A2, S] =
        ZIO.succeedLazy { (s, cont, f) =>
          self.foldUnwrap[R2, E2, A2, S].flatMap { foldLeft =>
            foldLeft(s, cont, f).flatMap { s =>
              if (!cont(s)) ZManaged.succeed(s)
              else
                ZManaged.unwrap {
                  other.fold[R2, E2, A2, S].map(foldRight => foldRight(s, cont, f))
                }
            }
          }
        }
    }
```

I guess that's decent enough!

Reply:
Another possibility is to use `ZManaged` directly in the `Fold` wrapper type. That could work too.